File: Zig/q_rsqrt.zig
pub fn q_rsqrt(i: f64) f64 {
    const threehalfs: f64 = 1.5;
    const half_i: f64 = 0.5 * i;
    const i_bits: u64 = 0x5fe6a09e667f3bc8 - (@bitCast(u64, i) >> 1);
    var out: f64 = @bitCast(f64, i_bits);

    out = out * (threehalfs - half_i * out * out);
    out = out * (threehalfs - half_i * out * out);
    out = out * (threehalfs - half_i * out * out);

    return out;
}

pub fn main() !void {
    const stdout = @import("std").io.getStdOut().writer();
    const testbits: u64 = 0x69420abcdef69420;
    var testcase: f64 = 0;
    testcase = @bitCast(f64, testbits);

    try stdout.print("{}\n", .{q_rsqrt(testcase)});
    try stdout.print("{}\n", .{q_rsqrt(0.5)});
}

File: C/q_rsqrt.c
#include <stdio.h> // so that I can print the values, q_rsqrt itself is not reliant on the C standard library.

double q_rsqrt(double i) {
    long long i_bits;
    double half_i, out;
    const double threehalfs = 1.5;

    half_i = i * 0.5;
    out = i;
    i_bits = * ( long long * ) &out;
    i_bits = 0x5fe6a09e667f3bc8 - ( i_bits >> 1 );
    out = * ( double * ) &i_bits;
    out = out * ( threehalfs - ( half_i * out * out ) ); // Newtonian approximation, lower two optional
    out = out * ( threehalfs - ( half_i * out * out ) );
    out = out * ( threehalfs - ( half_i * out * out ) );

    return out;
}

int main() {
    long long testbits = 0x69420abcdef69420;
    double testcase = * ( double * ) &testbits;

    printf("%.16e\n", q_rsqrt(testcase));
    printf("%.16e\n", q_rsqrt(0.5));

    return 0;
}
